analyses/molecular-subtyping-EPN/03-subgrouping_samples.py

```diff
@@ -0,0 +1,131 @@
+# This script takes in the final EPN table with all the columns (in results folder  results/EPN_all.tsv), 
+#	sets some rules and thresholds so samples can be  subgrouped
+#	and adds the sub group name in the last "subgroup" column
+
+## python 03-subgrouping_samples.py 
+#	--final_table results/EPN_all_data.tsv 
+#	--subgroup_table results/EPN_all_data_withsubgroup.tsv 
+#	--temp_folder_name subgrouping_samples   ## This folder will keep all the histograms to view  the outliers but can  be deleted  after review
+
+
+## Importing modules to be used
+import pandas as pd
+import numpy as np 
+import sys
+import argparse
+from matplotlib import pyplot as plt
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--final_table', required = True,
+                    help = 'path to the notebook with all EPN columns')
+parser.add_argument('--subgroup_table', required = True,
+		    help = 'output table  with subgroup')
+parser.add_argument('--temp_folder_name', required = True,
+                    help = 'outout table  with subgroup')
+args = parser.parse_args()
+
+# REading in input table  
+EPN_final = pd.read_csv(args.final_table, sep="\t")
+# Adding a column with empty strings so subgroup name can be added
+EPN_final["subgroup"] = ""
+
+# File to write all the data to
+outfile = open(args.subgroup_table, "w")
+
+# This function takes ina  row, value in the column from EPN_final table, 
+#	subgroupname that  should be returned if threshold is met, 
+#	value of the  threshold that  should  be higher in the row
+# It returns the subgroupname(which will be added to the final column)
+#	Before returning the subgroup name, it also checks if there is a subgroupname and
+#	appends to it  with a "," separator  
+def subgroup_func(row, columnname, subgroupname, value):
```

Comment:
I'm going to suggest a bit of a revision to this function to make your later code a bit simpler. The idea is to bind all of the `columname` & `value` pairs into a list of tuples so that we can pass in all of the tests for a single subgroup at once. This will allow you to make one call to `EPN_final.apply()` for each subgroup you want to type, and make it clearer what the tests being applied are.

For `ST_EPN_RELA`, that tuple might looks like this:

```python
st_epn_rela_tests = [("C11orf95--RELA", 0), 
                     ("LTBP3--RELA", 0),
                     ("9p_loss", 0),
                     ("9q_loss", 0)]
```

To use that, we could modify `subgroup_func()` as follows:

```python
def subgroup_func(row, subgroupname, column_values):
    current_subgroup = row["subgroup"]
    for columnname, value in column_values:
        if row[columnname] > value:
            if current_subgroup == '':
                current_subgroup = subgroupname
            elif subgroupname in current_subgroup.split(","):
                pass
            else:
                current_subgroup = current_subgroup + "," + subgroupname

    return(current_subgroup)
```

Then we should be able to call the whole subtyping for `ST_EPN_RELA` with
```python
EPN_final["subgroup"] = EPN_final.apply(subgroup_func,
                                        axis=1,
                                        subgroupname="ST_EPN_RELA",
                                        column_values=st_epn_rela_tests)
```